### ISSUE TITLE
A problem about latest global checkpoints

### ISSUE THREAD

**Issue by Philosober** [mllam/neural-lam#165]
Hi,

I’m using the latest global weather prediction checkpoints you provided in the `prob_model_global` branch. I followed the instructions in the README to run inference using the checkpoints from [this link](https://drive.google.com/drive/folders/1jkxpvrqiTgK9cw5-y2j5FZhLw6493el3).

However, the evaluation results I obtained differ significantly from those reported in the paper. I also noticed that the link has recently been updated, so I’m unsure how to proceed.

Could you kindly advise on how to reproduce the correct results?

Thank you very much for your help!

**Comment by joeloskarsson** [comment]
Hi, happy to help sort it out. Could you explain a bit more in detail what evaluation you tried running? The fact that the checkpoint links were updated should be unrelated, as I just moved the same files from onedrive to google drive.

Now it was a while since I touched the global branch, but the general sequence to run some evaluation for the global model should be:

1. Get the ERA5 data: https://github.com/mllam/neural-lam/tree/prob_model_global?tab=readme-ov-file#data
2. Run all pre-processing: https://github.com/mllam/neural-lam/tree/prob_model_global?tab=readme-ov-file#pre-processing
3. Download the checkpoints: https://github.com/mllam/neural-lam/tree/prob_model_global?tab=readme-ov-file#loading-pre-trained-models
4. Run your evaluation. An example eval script, that should match the main results from the paper, can be found at https://github.com/mllam/neural-lam/blob/prob_model_global/eval_scripts/graph_efm.sh.

Let me know where in the process things go wrong and I'll try to help figuring it out!

**Comment by Philosober** [comment]
Hi,

Thank you for your quick response. I followed the [README](https://github.com/mllam/neural-lam/tree/prob_model_global), which matches exactly what you described. I believe the issue might be related to the checkpoint link.

Could you please check the global weather prediction checkpoints for all graph-based models?

Initially, I used the GraphCast checkpoint for inference, but after noticing discrepancies from the paper results, I tried other checkpoints as well. Unfortunately, all the results still differ significantly from those reported in the paper.

I hope this information helps identify the root cause.

**Comment by joeloskarsson** [comment]
So for GraphCast for example, could you share the exact command you ran? Was it exactly https://github.com/mllam/neural-lam/blob/prob_model_global/eval_scripts/graphcast.sh? Also, what specifically makes you think that the checkpoint link is the issue? Is there something I am missing?

Is it possible to share some concrete example where you see a different result than in the paper? Is this in terms of how forecasts look, or e.g. some metric values?

**Comment by Philosober** [comment]
Yes! It was exactly same as [your link](https://github.com/mllam/neural-lam/blob/prob_model_global/eval_scripts/graphcast.sh). I just guess because the data downloading and preprocessing is very clear as following (graphcast):
1. downloading from weatherbenchv2
2. creating graph `python create_global_mesh.py --dataset global_era5 --graph global_multiscale --splits 4`
3. creating global grid features `python global_grid_features.py --dataset global_era5`
4. creating global forcing `python create_global_forcing.py --dataset global_era5`
5. creating parameter weights `python create_parameter_weights.py --dataset global_era5`

Then I download the checkpoint from your link, and results are as follows:
[test_rmse.pdf](https://github.com/user-attachments/files/23291111/test_rmse.pdf). Again, thanks for your quick reply!

**Comment by Philosober** [comment]
Information updating: I train graphcast from scratch, and find the current result is better than the checkpoint.

**Comment by joeloskarsson** [comment]
Thanks for giving all the details. Yeah, those rmse values are completely off, almost looks like an untrained model. I would almost think that the code somehow doesn't use the checkpoint, but that does not really make sense. I have tested that this should work. Maybe there is indeed something up with the checkpoint files since when I copied them over. 

Just to make sure there is nothing going wrong with the graphs, could you try downloading and using the graph from https://drive.google.com/drive/folders/1SMgEQbMO8J-r32BUk8t4pbr7pQCe0mkk?usp=sharing instead of the one you generated?

I will have to dig a bit deeper into this and test it out myself. Sadly I'm a bit busy this week as I just started a new position. But I will look into it as soon as i have time.

**Comment by Philosober** [comment]
Thanks for your ongoing replies. I tried again on your provided graph, but the result ([test_rmse.pdf](https://github.com/user-attachments/files/23354038/test_rmse.pdf)) is still as the same. It may take you some time to help find the problem out as soon as you have time.

**Comment by Philosober** [comment]
Forgot to share the important information. All the checkpoints' results including Graph-FM and Graph-EFM also have the same problem as GraphCast.

**Comment by Philosober** [comment]
For curiosity, I dive into the graphs you provided. I find your graphs are different from graphs I generated. Specifically, `m2g_edge_index.pt`'s 60 edges are different (Only This), which results in the difference of `m2g_features.pt`. The instruction I used is: `python create_global_mesh.py --dataset global_era5 --graph global_multiscale --splits 4`.

**Comment by joeloskarsson** [comment]
Hmm, this is odd, but could be useful in pointing us in the right direction. The only part I can not control for on your side is the actual ERA5 data. If there are differences there (in particular in grid node positions) that could also result in a different graph. You have [this exact version](https://console.cloud.google.com/storage/browser/weatherbench2/datasets/era5/1959-2023_01_10-6h-240x121_equiangular_with_poles_conservative.zarr?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false) of ERA5 right?

It is however also the case that you can end up with an equivalent graph, even if the edge_index tensor differs (you can re-arrange the order of edges).

**Comment by Philosober** [comment]
The ERA5's version is right. This is worse. I have no idea how to figure it out......

**Comment by Philosober** [comment]
After diving into the `create_graph_mesh.py`, I found a problem. Considering that the grid's structure is (lon, lat), `senders=g2m_edge_index[0, :]` may be incorrect because the `g2m_edge_index` is not revised by the `grid_index_map`.
```
    g2m_edge_index = np.stack(g2m_grid_mesh_indices, axis=0)
    g2m_edge_index_torch = torch.tensor(g2m_edge_index, dtype=torch.long)
    # Grid index fix
    g2m_edge_index_torch[0] = grid_index_map[g2m_edge_index_torch[0]]

    # Only care about edge features here
    _, _, g2m_features = gc_mu.get_bipartite_graph_spatial_features(
        senders_node_lat=grid_lat_lon_flat[:, 0],
        senders_node_lon=grid_lat_lon_flat[:, 1],
        # senders=g2m_edge_index[0, :],
        senders=grid_index_map[g2m_edge_index[0, :]],   # revise
        receivers_node_lat=grid_con_mesh_lat_lon[:, 0],
        receivers_node_lon=grid_con_mesh_lat_lon[:, 1],
        receivers=g2m_edge_index[1, :],
        **GC_SPATIAL_FEATURES_KWARGS,
    )
    g2m_features_torch = torch.tensor(g2m_features, dtype=torch.float32)
```

**Comment by joeloskarsson** [comment]
Not sure if that is a problem, `g2m_edge_index` just contains node indices (unrelated to lon/lat ordering). The first row there should be the sender indices. The re-mapping of indexes is only to make the graphcast code indexing compatible with our code, so it should not be needed in functions that call the graphcast implementation.

Sorry for the lack of updates from my side. I promise that I have not forgot about this and will look into it asap (I should have some time for this this weekend). Since I know that I have reproduced results in this setup before, I am wondering if the issues could relate to some package version differences, or otherwise be data-related. But I need to dig into it and see if the checkpoints work correctly on my side.

**Comment by Philosober** [comment]
Let me further explain. The `g2m_edge_index`'s sequence is lat/lon because it is from the graphcast's function. However, the `grid_lat_lon_flat`'s sequence is lon/lat, right? Therefore, you revise it by the `grid_index_map`. OK, the problem occurs in calling `gc_mu.get_bipartite_graph_spatial_features(....)` function. The `grid_lat_lon_flat` do not match the `g2m_edge_index` from the sequence perspective. `# revise` is the line I revised.
Thank you very much for your continuing response! I will still find the problem as well.

**Comment by joeloskarsson** [comment]
I started investigating this now, and I don't think that the checkpoint files themselves are the problem. I've been looking at the t2m prediction, which starts out with a really high error already at time step 1. It looks almost like the model is predicting day temperatures during night and the other way around, as if the model got the wrong time of day or TOA radiation. So current hypothesis is that it would relate to the forcing somehow. Digging through some old files, but still nothing conclusive.

**Comment by joeloskarsson** [comment]
Ok I think I confirmed that the problem relates to the forcing. It might be that there has been updates to the graphcast code (which the forcing generation uses) that impacts the exact forcing values created. I managed to get forecasts that look correct by using the static and forcing data found here: https://drive.google.com/drive/folders/1ag2mZLYQiHWg-McFNRR92xSoria-Zs73?usp=sharing You can download that and give it a try (and let me know if that fixes it also for you, would be super helpful).

**Comment by joeloskarsson** [comment]
I went ahead and ran a full test set eval with the GraphCast* model using the forcing data linked above. This seems to solve the problem completely, and I get results that are identical to the paper up to some decimals (I ran this on multi-GPU as well, so expecting some differences, but they turned out negligable). Attaching the RMSEs from that evaluation here:

<img width="2500" height="2500" alt="Image" src="https://github.com/user-attachments/assets/ecf559f5-a6ce-429a-96c1-694210450e96" />

I still need to look into exactly why the TOA radiation is different in this forcing data I computed earlier, and the one we have computed now. But from initial observations it should just be some small misalignment, which however the pre-trained checkpoints do not play well with. So you should be able to 1) Use the pre-trained checkpoint with the forcing data I shared, 2) succesfully train new models from scratch using newly computed forcing data.

**Comment by Philosober** [comment]
Thanks bro! I got a terrible flu recently. I swear I don't forget this thing, and I already feel better today. Then I will continue diving into this problem and have a try following your advice.

**Comment by joeloskarsson** [comment]
Finally figured out the full problem here 😄 So the issue was not just the TOA radiation field, but really the ordering of the different forcing fields. In my "old" forcing zarr, that was used to train the checkpoints, the forcing fields are sorted by name as 
```
array(['cos_day_progress', 'cos_year_progress', 'sin_day_progress', 'sin_year_progress', 'toa_incident_radiation'], dtype=object)
```
but when running the forcing generation now, the forcing fields are sorted according to the ordering they are created in the script:
```
array(['toa_incident_radiation', 'sin_year_progress', 'cos_year_progress', 'sin_day_progress', 'cos_day_progress'], dtype=object)
```
This of course causes big issues when using the pre-trained checkpoints. They would for example interpret the `cos_day_progress` field in the newly generated forcing as TOA radiation. 

I can't really find when this was changed (I don't think I ever sorted this manually 🤔). Perhaps it was some change in `xarray` that caused this discrepancy. The different forcing fields are put in a dict, so that should probably have been an ordered dict, but python dicts have been guaranteed to retain insertion order for some time now, so I'm a bit confused. Anyhow, no need to continue investigating now that we know the problem. 

I pushed a simple fix here: https://github.com/mllam/neural-lam/commit/1a6e511cbca63835114f3a90a0b1ff60cb453ac1 Now the forcing generation script should generate an ordering of fields compatible with the checkpoints. I also tested this. So you should just be able to re-make the forcing zarr. I will close this for now, and just let me know if you encounter more issues, then we can re-open it 😄 And thanks again for making me aware of this so I could fix it!

**Comment by Philosober** [comment]
Bro, thanks so much for your help. I ran your new `create_forcing.py` and got the same results you sent. I caught pneumonia last month and had to be hospitalized, which took a while to recover from. I’m really sorry for the late reply (I was in the hospital at the time🤦). It's great connecting with you, and I'm looking forward to your future work in deep learning!

**Comment by joeloskarsson** [comment]
Sucks to hear about the sickness, but good that you recovered! Great that the fix seems to work for you :)
